### PR TITLE
Backend/refactor/ #BKN-2787 apply finalizer refactor

### PR DESCRIPTION
**Pull request by roman-bodavskiy**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates

### Additional Changes

- [ ] This PR modifies the database schema (database migration added)
- [ ] This PR modifies dhall configs/environment variables

## Motivation and Context
https://juspay.atlassian.net/browse/BKN-2787

Refactored this PR by Pavel Zarubin
https://github.com/nammayatri/shared-kernel/pull/26
In my solution we don't need to make `SqlDB` monad type polymorphic and change all existing queries.
Instead I use existential quantification:
```
data SqlDBEnv = forall m.
  Finalize m =>
  SqlDBEnv
  { currentTime :: UTCTime,
    actions :: m ()
  }
```

## How did you test it?
Tested locally. Tested transaction inside of transaction. All finalizer actions worked as expected.

## Checklist

- [ ] I formatted the code and addressed linter errors `./dev/format-all-files.sh`
- [ ] I reviewed submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: lib/mobility-core/src/Kernel/Storage/Esqueleto/Transactionable.hs [hunk 1]
```diff
@@ -55,48 +68,99 @@ instance {-# OVERLAPPING #-} Transactionable' SqlDB m => Transactionable' SqlDB
...
+  let m = mkAction (finalizeImpl @m)
+  (result, env) <- runTransactionImpl dbEnv m
+  applyFinalizeImpl @m env
+  pure result
+
+-- | This function should not be used outside of Transactionable module!
+finalizeImpl :: forall m. Finalize m => m () -> SqlDB ()
+finalizeImpl someAction = do
+  SqlDBEnv {..} <- get
+  let mbPrevActions = cast @_ @(m ()) actions
```
Comment (DonDokidon) [inline lib/mobility-core/src/Kernel/Storage/Esqueleto/Transactionable.hs:94]: Is it possible to remove cast & Typeable instance from the whole thing? We won't use it without runTransactionF and monad  will be decided with first finalize call anyway

**Review by DonDokidon — CHANGES_REQUESTED** [review]

File: lib/mobility-core/src/Kernel/Storage/Esqueleto/Transactionable.hs (hunk 1 above)
Comment (roman-bodavskiy) [reply lib/mobility-core/src/Kernel/Storage/Esqueleto/Transactionable.hs:94]: > Is it possible to remove cast & Typeable instance from the whole thing? We won't use it without runTransactionF and monad will be decided from the start

I don't know how its possible, If you know please add your variant here. In previous solution by @pavelzarubin he didn';t use cast & Typeable instance. 
https://github.com/nammayatri/nammayatri/pull/67
But therefore he needed to make `SqlDB` type polimorphic (`SqlDB m a`). It required a lot of changes all over the code, so my solution I like better.

Comment (roman-bodavskiy) [reply lib/mobility-core/src/Kernel/Storage/Esqueleto/Transactionable.hs:94]: Another my solution was to create some local type synonyms in each app, like:

`type SqlDB a = Esq.SqlDB Flow a`

but its not working with polymorphic code, when we don't know exact type of `m`

**Review by DonDokidon — APPROVED** [review]

**Review by pavelzarubin — APPROVED** [review]

**Comment by roman-bodavskiy** [comment]
using Beam instead of Esqueleto now
transactions not required

**Pull request closed without merging**